xapian-applications/omega/clickmodel/tests/sdbntest.cc

```diff
@@ -46,10 +46,15 @@ struct sessions_testcase {
     const sessions_items sessions;
 };
 
-char *p = getenv("srcdir");
-string sample_log1 = string(p) + "/clickmodel/testdata/test1.log";
-string sample_log2 = string(p) + "/clickmodel/testdata/test2.log";
-string sample_log3 = string(p) + "/clickmodel/testdata/test3.log";
+static string get_srcdir() {
+    char *p = getenv("srcdir");
+    if (!p) return ".";
+    return string(p);
+}
+
+string sample_log1 = get_srcdir() + "/clickmodel/testdata/test1.log";
+string sample_log2 = get_srcdir() + "/clickmodel/testdata/test2.log";
+string sample_log3 = get_srcdir() + "/clickmodel/testdata/test3.log";
```

Comment:
I wrote:

> It would also be much better to do this initialisation inside a function rather than as global static objects. It's hard to usefully handle exceptions or report errors during global static object initialisation - a good rule of thumb is to avoid it completely where you can, and keep it simple where you actually need it.

The change above isn't what I meant - I was suggesting to avoid unnecessarily doing work during global object initialisation, not just moving the `getenv()` call into its own function which still gets called during global object initialisation

What we're trying to avoid is doing any complex work at runtime before we enter `main()` - it's hard to usefully handle exceptions or report errors before `main()` - there's no enclosing scope to catch exceptions and I believe you can't even necessarily rely on things like `cout` working properly at this stage (because `cout` is also a global object, and the order of initialisation of global objects isn't specified, so your global object may get initialised before `cout` is initialised).

What I was proposing was to move this initialisation to the start of the `main()` function.

Also, your change means we're now calling `getenv("srcdir")`, etc multiple times - that'll be easier to avoid that if you do this at the start of `main()`, because you can easily have a temporary `srcdir` variable.

Reply:
Alright, I'll move the filepath initialisations inside the main function along with the `sessions_tests[]` struct array. Since, `sessions_tests[]` was declared globally earlier which uses the filepaths `sample_log1` and `sample_log2`, I had to make those initialisations global as well.